### PR TITLE
[training, recipe] feat: Add unified run_recipe.py and Gemma3 VL examples

### PR DESCRIPTION
**Pull request by yaoyu-33**
# Update Train Scripts, Recipes, and Add Gemma3 VL Examples

## Summary

This PR updates the training infrastructure and adds comprehensive Gemma 3 VL example scripts. Key changes include:

1. **Unified training script** - Replaced separate `pretrain_decoder.py` and `finetune_decoder.py` with a single `run_recipe.py` that works for all training modes
2. **VLM recipe defaults** - Updated VLM recipes to default to `"hf"` dataset type instead of `"mock"` for more realistic out-of-box experience
3. **Gemma 3 VL examples** - Added complete example scripts for checkpoint conversion, SFT, PEFT, and inference

### Training Scripts (`scripts/training/`)

- **Added**: `run_recipe.py` - Generic training script for GPT-based models that:
  - Dynamically loads recipes from `megatron.bridge.recipes`
  - Supports CLI overrides with dot notation (e.g., `train.train_iters=5000`)
  - Auto-infers training mode from recipe name (`pretrain`/`finetune`)
  - Supports multiple step functions: `gpt_step`, `vlm_step`, `llava_step`
  - Supports PEFT schemes via `--peft_scheme lora`

- **Removed**: `pretrain_decoder.py` and `finetune_decoder.py` (consolidated into `run_recipe.py`)
- **Removed**: `conf/template_overrides.yaml` (no longer needed)

- **Updated**: `README.md` - Streamlined documentation for the new unified script
- **Updated**: `launch_with_nemo_run.py` and `launch_with_sbatch.sh` - Minor improvements

### VLM Recipes

Updated default dataset type from `"mock"` to `"hf"` for better out-of-box experience:
- `src/megatron/bridge/recipes/gemma3_vl/gemma3_vl.py`
- `src/megatron/bridge/recipes/glm_vl/glm_45v.py`
- `src/megatron/bridge/recipes/ministral3/ministral3.py`
- `src/megatron/bridge/recipes/qwen_vl/qwen25_vl.py`
- `src/megatron/bridge/recipes/qwen_vl/qwen3_vl.py`

### Gemma 3 VL Examples (`examples/models/vlm/gemma3_vl/`)

Added complete example directory with:
- `README.md` - Overview and links to each script
- `conversion.sh` - HF ↔ Megatron checkpoint conversion commands
- `sft.sh` - Full-parameter supervised fine-tuning with TP/PP variants
- `peft.sh` - LoRA fine-tuning with TP/PP variants  
- `inference.sh` - Inference with HF, imported, and exported checkpoints

Example usage from new scripts:
```bash
# SFT training
uv run python -m torch.distributed.run --nproc_per_node=8 scripts/training/run_recipe.py \
    --recipe gemma3_vl_4b_finetune_config \
    --step_func vlm_step \
    checkpoint.pretrained_checkpoint=/models/gemma-3-4b-it \
    train.train_iters=50

# LoRA training
uv run python -m torch.distributed.run --nproc_per_node=8 scripts/training/run_recipe.py \
    --recipe gemma3_vl_4b_finetune_config \
    --step_func vlm_step \
    --peft_scheme lora \
    checkpoint.pretrained_checkpoint=/models/gemma-3-4b-it
```

## Migration Notes

If you were using the old training scripts:

**Before:**
```bash
torchrun --nproc_per_node=8 scripts/training/pretrain_decoder.py \
    --recipe llama32_1b_pretrain_config
```

**After:**
```bash
torchrun --nproc_per_node=8 scripts/training/run_recipe.py \
    --recipe llama32_1b_pretrain_config
```

The mode is now automatically inferred from the recipe name. If your recipe doesn't include `pretrain` or `finetune`, pass `--mode` explicitly.

## Testing

- [ ] Verified Gemma 3 VL SFT training
- [ ] Verified Gemma 3 VL PEFT training
- [ ] Verified checkpoint conversion round-trip
- [ ] Verified inference works with all checkpoint types

## Checklist

- [x] Updated training scripts to unified `run_recipe.py`
- [x] Updated VLM recipes default dataset type
- [x] Added Gemma 3 VL example scripts
- [x] Updated README documentation

## Summary by CodeRabbit

* **New Features**
  * Added Gemma 3 Vision Language Model examples with setup guides and training/inference scripts.
  * Introduced a unified training launcher for simplified model configuration and execution.

* **Documentation**
  * Updated training workflow documentation with new launcher usage and step function controls.

* **Improvements**
  * Changed default dataset selection for Vision Language models from mock to real datasets.
  * Removed legacy configuration templates and deprecated training entry points.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: scripts/training/run_recipe.py [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+
+def load_forward_step(step_type: str):
+    """Load forward_step function based on the requested step type."""
+    step_key = step_type.lower()
+    if step_key not in STEP_MODULES:
+        raise ValueError(f"Unknown step type: {step_type}. Choose from: {', '.join(STEP_MODULES)}")
+    module = importlib.import_module(STEP_MODULES[step_key])
+    if not hasattr(module, "forward_step"):
+        raise AttributeError(f"{STEP_MODULES[step_key]} does not define forward_step")
```
Comment (ananthsub) [inline scripts/training/run_recipe.py]: why does this need to use importlib considering there's a fixed, known set of step funcs to run? is it save on import dependencies?

Comment (yaoyu-33) [reply scripts/training/run_recipe.py]: yeah, I can update

File: scripts/training/launch_with_sbatch.sh [hunk 2]
```diff
@@ -44,9 +44,9 @@
 # ==============================================================================
 
 # Training script to run
-TRAINING_SCRIPT="pretrain_decoder.py"
+TRAINING_SCRIPT="run_recipe.py"
 # Options:
-# TRAINING_SCRIPT="finetune_gpt.py"
+# TRAINING_SCRIPT="run_recipe.py"
 # TRAINING_SCRIPT="pretrain_vlm.py"  # For VLM models
 # TRAINING_SCRIPT="finetune_vlm.py"  # For VLM finetuning
```
Comment (ananthsub) [inline scripts/training/launch_with_sbatch.sh:51]: remove extra comments - none of these are relevant now

File: scripts/training/README.md [hunk 3]
```diff
@@ -6,11 +6,9 @@ Generic launcher and training scripts that work with any GPT-based model family
 
 These scripts provide a generic interface for training GPT-based models in Megatron Bridge:
 
-- `pretrain_decoder.py` - Generic pretraining for GPT- and Mamba-based models.
-- `finetune_decoder.py` - Generic finetuning for GPT- and Mamba-based models.
+- `run_recipe.py` - Generic pretraining/finetuning for GPT- and Mamba-based models.
```
Comment (ananthsub) [inline scripts/training/README.md:9]: also applies for VLMs now

File: scripts/training/launch_with_sbatch.sh [hunk 4]
```diff
@@ -58,9 +58,8 @@ RECIPE="llama32_1b_pretrain_config"
 # RECIPE="llama3_8b_pretrain_config"
 # RECIPE="qwen25_vl_pretrain_config"  # For VLM models
 
-# Optional: YAML config file
-CONFIG_FILE=""
-# CONFIG_FILE="conf/my_custom_config.yaml"
+# Forward step type (gpt or vlm)
+STEP_TYPE="gpt"
```
Comment (ananthsub) [inline scripts/training/launch_with_sbatch.sh:62]: this doesn't match `gpt_step` in the step func dictionary from run_recipe.py

File: scripts/training/launch_with_nemo_run.py [hunk 5]
```diff
@@ -147,7 +146,7 @@ def parse_args() -> tuple[argparse.Namespace, list[str]]:
         "--script",
         type=str,
         required=True,
-        help="Training script to run (e.g., pretrain_decoder.py, finetune_decoder.py, pretrain_vlm.py)",
+        help="Training script to run (e.g., run_recipe.py, pretrain_vlm.py, finetune_vlm.py)",
```
Comment (ananthsub) [inline scripts/training/launch_with_nemo_run.py:149]: should this default to using run_recipe.py now that the others aren't needed?

**Review by ko3n1g — APPROVED** [review]
Agreed offline that tests will come afterwards
